File: csharp/demo/demo08b-return-value.cs
/*
 * GETTING RETURNED VALUES FROM THREADS
 * Using AutoResetEvent to notify the thread which is waiting for returned values
 */
using System;
using System.Threading;



class Demo08B : IRunnable
{
    private AutoResetEvent re;
    private int result;


    public void run()
    {
        re = new AutoResetEvent(false);

        new Thread(() => doubleValue(5)).Start();

        // Wait until we receive a notification from re
        re.WaitOne();

        Console.WriteLine(result);
    }


    private void doubleValue(int value)
    {
        result = value * 2;

        Thread.Sleep(2000);
        re.Set(); // Notify the threads that are waiting for re

        Thread.Sleep(2000);
        Console.WriteLine("This thread is exiting");
    }
}

File: python/demo08b_return_value.py
'''
GETTING RETURNED VALUES FROM THREADS
'''

import threading



def double_value(result: dict, name: str, value):
    result[name] = value * 2



res = {}

th_foo = threading.Thread( target=double_value, args=(res, 'foo', 5) )
th_bar = threading.Thread( target=double_value, args=(res, 'bar', 80) )

th_foo.start()
th_bar.start()

# Wait until th_foo and th_far finish
th_foo.join()
th_bar.join()

print(res['foo'])
print(res['bar'])
